persistence-api/src/main/java/io/stargate/db/datastore/PersistenceBackedDataStore.java

```diff
@@ -1,81 +1,178 @@
-        throw new ProtocolException("Unsupported protocol version: " + version);
-      case V3:
-        return ProtocolVersion.V3;
-      case V4:
-        return ProtocolVersion.V4;
-      case V5:
-        return ProtocolVersion.V5;
-      default:
-        throw new AssertionError("Unhandled protocol version: " + version);
-    }
+    this.options = options;
+    this.valueCodec = new Codec(parameters().protocolVersion(), connection.persistence());
+  }
+
+  private Parameters parameters() {
+    return options.defaultParameters();
   }
 
   @Override
-  public CompletableFuture<ResultSet> query(
-      String queryString, UnaryOperator<Parameters> parametersModifier, Object... values) {
-    return prepare(queryString).thenCompose(p -> p.execute(parametersModifier, values));
+  public Codec valueCodec() {
+    return valueCodec;
+  }
+
+  private Statement toPersistenceStatement(BoundQuery query) {
+    List<TypedValue> values = query.values();
+    List<ByteBuffer> buffers = new ArrayList<>(values.size());
+    for (TypedValue value : values) {
+      buffers.add(value.bytes());
+    }
+    Query<?> boundedQuery = query.bounded().query();
+    return boundedQuery.preparedId().isPresent()
+        ? new BoundStatement(boundedQuery.preparedId().get(), buffers, null)
+        : new SimpleStatement(boundedQuery.queryStringForPreparation(), buffers);
+  }
+
+  private void validateExecuteParameters(Parameters executeParameters) {
+    if (parameters() == executeParameters) {
+      return;
+    }
+
+    Preconditions.checkArgument(
+        !executeParameters.skipMetadataInResult(),
+        "Invalid execution parameters: you should not set 'skipMetadataInResult', this"
+            + "is handled internally by DataStore.");
+    Preconditions.checkArgument(
+        executeParameters.protocolVersion() == parameters().protocolVersion(),
+        "Invalid execution parameters: cannot modify the protocol version for execution "
+            + "(the DataStore version %s != %s, the execution parameters version).",
+        parameters().protocolVersion(),
+        executeParameters.protocolVersion());
   }
 
   @Override
-  public CompletableFuture<PreparedStatement> prepare(String queryString) {
+  public <B extends BoundQuery> CompletableFuture<Query<B>> prepare(Query<B> query) {
     return connection
-        .prepare(queryString, parameters)
-        .thenApply(
-            prepared ->
-                new PersistenceBackedPreparedStatement(
-                    connection, parameters, new PreparedInfo(prepared), queryString));
+        .prepare(query.queryStringForPreparation(), parameters())
+        .thenApply(prepared -> query.withPreparedId(prepared.statementId));
   }
 
   @Override
-  public CompletableFuture<ResultSet> batch(List<String> queries) {
+  public CompletableFuture<ResultSet> execute(
+      BoundQuery query, UnaryOperator<Parameters> parametersModifier) {
     long queryStartNanos = System.nanoTime();
```

Comment:
What do you mean by "it"? The `parametersModifier` parameter? If so, it's just the most generic way to modify query parameters, but there is method overloads, so it's not passed to every call in practice.

Reply:
sorry I meant the `long queryStartNanos = System.nanoTime();`

Reply:
Oh, my bad. That is time is passed all the way down to C*, which uses it to enforce its timeouts (and possibly some metrics, I don't remember exactly). Pushing its generation down would change a bit what timeouts covers exactly for users and it could be debated what option is better, though I personally feel generating it early (as is done now) is probably better. In any case, I'll remark that I'm preserving the existing behavior here.